proposals/lip-0052.md

```diff
@@ -362,712 +379,1832 @@ The module provides the following commands to modify the NFT store.
+        },
+        topics = [module, nftID]
+    )
+
+    def emitFailedLockEvent(
+        module: Module,
+        nftID: NFTID,
+        result: uint32
+    ) -> None:
+
+        emitPersistentEvent(
+            module = MODULE_NAME_NFT,
+            name = EVENT_NAME_LOCK,
+            data = {
+                "module": module,
+                "nftID": nftID,
+                "result": result
+            },
+            topics = [module, nftID]
+        )
+```
+
+#### unlock
 
-    create an entry in the collection substore with
-        storeKey = uint32be(collection)
-        storeValue = {"nextAvailableIndex": 0) serialized using collectionStoreSchema
+This function unlocks an NFT that was previously locked to a module.
 
-    if collection >= nextAvailableCollection:
-        nextAvailableCollection = collection + 1
-    return collection
+```python
+def unlock(
+    module: Module,
+    nftID: NFTID
+) -> None:
+
+    if NFTStore(nftID) does not exist:
+        emit FailedUnlockEvent(module, nftID, RESULT_NFT_DOES_NOT_EXIST)
+        raise Exception("NFT store entry does not exist.")
+
+    if getLockingModule(nftID) == NFT_NOT_LOCKED:
+        emit FailedUnlockEvent(module, nftID, RESULT_NFT_NOT_LOCKED)
+        raise Exception("NFT is not locked.")
+
+    if getLockingModule(nftID) != module:
+        emit FailedUnlockEvent(module, nftID, RESULT_UNAUTHORIZED_UNLOCK)
+        raise Exception("Unlocking NFT via module that didn't lock it.")
+
+    userStore(getNFTOwner(nftID), nftID).lockingModule = NFT_NOT_LOCKED
+
+    emitEvent(
+        module = MODULE_NAME_NFT,
+        name = EVENT_NAME_UNLOCK,
+        data = {
+            "module": module,
+            "nftID": nftID,
+            "result": RESULT_SUCCESSFUL
+        },
+        topics = [module, nftID]
+    )
+
+    def emitFailedUnlockEvent(
+        module: Module,
+        nftID: NFTID,
+        result: uint32
+    ) -> None:
+
+        emitPersistentEvent(
+            module = MODULE_NAME_NFT,
+            name = EVENT_NAME_UNLOCK,
+            data = {
+                "module": module,
+                "nftID": nftID,
+                "result": result
+            },
+            topics = [module, nftID]
+        )
 ```
 
-#### transfer
+#### setAttributes
 
-This function transfers ownership of NFTs within one chain.
+This function modifies the attributes of an NFT.
 
 ```python
-transfer(senderAddress, recipientAddress, nftID):
-    nftID = canonicalNFTID(nftID)
-    if (NFTStore(nftID).lockingModuleID != NFT_NOT_LOCKED
-        or NFTStore(nftID).owner != senderAddress):   
-        transfer fails
+def setAttributes(
+    module: Module,
+    nftID: NFTID,
+    attributes: Attributes
+) -> None:
+
+    if NFTStore(nftID) does not exist:
+        emitFailedSetAttributesEvent(module, nftID, attributes, RESULT_NFT_DOES_NOT_EXIST)
+        raise Exception("NFT substore entry does not exist.")
+
+    if getChainID(nftID) != OWN_CHAIN_ID:
```

Comment:
Why are we limiting modifying attributes only to native chains? Isn't it an important use case to modify non-native NFTs? Maybe @MaximeGagnebin can chim in?

Reply:
Yes, I put it like that for new because the cross-chain modification of attributes is something I would like to discuss with you. Actually, the same goes with all modification of attributes, even on native chain, since we don't have implemented the distinction between 'fixed' and 'mutable' attributes. 

As I see it, 'fixed' attributes would be the ones set by the collectionID (or particular NFTs, to be decided) creator, and could not be mutated by anyone. The NFT creator would then be able to decide whether he allows for the existence of 'mutable' attributes that other modules (both on native and foreign chains) can then mutate.

In my view, something to be discussed on an STM before (potentially) implementing it in the protocol. So, this setAttributes function is really unfinished business for now.

Reply:
can be resolved